edb/server/compiler/sql.py

```diff
@@ -109,9 +111,60 @@ def _try(q: str) -> List[dbstate.SQLQueryUnit]:
             raise original_err
 
 
+def _build_constant_extraction_map(
+    src: pgast.Base,
+    out: pgast.Base,
+) -> pg_codegen.BaseTranslationData:
+    """Traverse two ASTs in parallel and build a source map between them.
+
+    The ASTs should *mostly* line up. When they don't, that is
+    considered a leaf.
+
+    This is used to translate SQL spans reported on a normalized query
+    to ones that make sense on the pre-normalization version.
+
+    Note that we only use this map for errors reported during the
+    "parse" phase, so we don't need to worry about it being reused
+    with different constants.
+    """
+    tdata = pg_codegen.BaseTranslationData(
+        source_start=src.span.start if src.span else 0,
+        # HACK: I don't know why, but this - 1 helps a lot.
+        output_start=out.span.start - 1 if out.span else 0,
+    )
+    if type(src) != type(out):
+        return tdata
```

Comment:
Maybe also check that one of the types is a `ParamRef`? Just for the case when in the future we do more aggressive normalization.

Reply:
I think we'll still probably want to map the spans then, though?